qa/smoke-test-http/src/test/java/org/elasticsearch/http/DanglingIndicesRestIT.java

```diff
@@ -0,0 +1,291 @@
+    }
+
+    /**
+     * Check that dangling indices can be imported.
+     */
+    public void testDanglingIndicesCanBeImported() throws Exception {
+        internalCluster().setBootstrapMasterNodeIndex(1);
+        internalCluster().startNodes(3, buildSettings(0));
+
+        createDanglingIndices(INDEX_NAME);
+
+        final RestClient restClient = getRestClient();
+
+        final List<String> danglingIndexIds = listDanglingIndexIds();
+        assertThat(danglingIndexIds, hasSize(1));
+
+        final Request importRequest = new Request("POST", "/_dangling/" + danglingIndexIds.get(0));
+        importRequest.addParameter("accept_data_loss", "true");
+        // Ensure this parameter is accepted
+        importRequest.addParameter("timeout", "20s");
+        final Response importResponse = restClient.performRequest(importRequest);
+        assertThat(importResponse.getStatusLine().getStatusCode(), equalTo(ACCEPTED.getStatus()));
+
+        final XContentTestUtils.JsonMapView mapView = createJsonMapView(importResponse.getEntity().getContent());
+        assertThat(mapView.get("accepted"), equalTo(true));
+
+        assertTrue("Expected dangling index " + INDEX_NAME + " to be recovered", indexExists(INDEX_NAME));
+    }
+
+    /**
+     * Check that dangling indices can be deleted. Since this requires that
+     * we add an entry to the index graveyard, the graveyard size must be
+     * greater than 1. To test deletes, we set the index graveyard size to
+     * 1, then create two indices and delete them both while one node in
+     * the cluster is stopped. The deletion of the second pushes the deletion
+     * of the first out of the graveyard. When the stopped node is resumed,
+     * only the second index will be found into the graveyard and the the
+     * other will be considered dangling, and can therefore be listed and
+     * deleted through the API
+     */
+    public void testDanglingIndicesCanBeDeleted() throws Exception {
+        internalCluster().setBootstrapMasterNodeIndex(1);
+        internalCluster().startNodes(3, buildSettings(1));
+
+        createDanglingIndices(INDEX_NAME, OTHER_INDEX_NAME);
+
+        final RestClient restClient = getRestClient();
+
+        final List<String> danglingIndexIds = listDanglingIndexIds();
+        assertThat(danglingIndexIds, hasSize(1));
+
+        final Request deleteRequest = new Request("DELETE", "/_dangling/" + danglingIndexIds.get(0));
+        deleteRequest.addParameter("accept_data_loss", "true");
+        // Ensure these parameters is accepted
+        deleteRequest.addParameter("timeout", "20s");
+        deleteRequest.addParameter("master_timeout", "20s");
+        final Response deleteResponse = restClient.performRequest(deleteRequest);
+        assertThat(deleteResponse.getStatusLine().getStatusCode(), equalTo(ACCEPTED.getStatus()));
+
+        assertBusy(() -> assertThat("Expected dangling index to be deleted", listDanglingIndexIds(), hasSize(0)));
```

Comment:
We can demonstrate that the dangling index was not properly deleted by adding this:

```
        createIndex("additional");
        deleteIndex("additional");
        assertThat(listDanglingIndexIds(), empty());
```

The creation/deletion of this extra index pushes the tombstone out of the graveyard which resurrects it.

(gruesome nomenclature isn't it? :grin:)

Reply:
Well that worked, in the sense that it showed that the index isn't being deleted off disk. I'll see if I can figure out how to delete it properly, but if you have any pointers, that'd be great.

Reply:
One idea is for `ClusterChangedEvent#indicesDeletedFromClusterState` to consider added tombstones as an indication that an index was deleted as well as removed index metadata.

Reply:
@DaveCTurner see commit d02c5ce3113. The tests now passes, but I'm really not sure about the changes. Can you please take a look?

Reply:
Yep, looks like what we want. Suggest using a `Set<Index>` rather than a list and a membership check but otherwise fine.

Reply:
I switched to a set in b94efdf802b.